velox/exec/NestedLoopJoinProbe.cpp

```diff
@@ -28,48 +54,64 @@ NestedLoopJoinProbe::NestedLoopJoinProbe(
           operatorId,
           joinNode->id(),
           "NestedLoopJoinProbe"),
-      outputBatchSize_{outputBatchRows()} {
+      outputBatchSize_{outputBatchRows()},
+      joinType_(joinNode->joinType()) {
   auto probeType = joinNode->sources()[0]->outputType();
-  for (auto i = 0; i < probeType->size(); ++i) {
-    auto name = probeType->nameOf(i);
-    auto outIndex = outputType_->getChildIdxIfExists(name);
-    if (outIndex.has_value()) {
-      identityProjections_.emplace_back(i, outIndex.value());
-    }
-  }
-
   auto buildType = joinNode->sources()[1]->outputType();
-  for (auto i = 0; i < outputType_->size(); ++i) {
-    auto tableChannel = buildType->getChildIdxIfExists(outputType_->nameOf(i));
-    if (tableChannel.has_value()) {
-      buildProjections_.emplace_back(tableChannel.value(), i);
-    }
+  identityProjections_ = extractProjections(probeType, outputType_);
+  buildProjections_ = extractProjections(buildType, outputType_);
+
+  if (joinNode->joinCondition() != nullptr) {
+    initializeFilter(
+        joinNode->joinCondition(),
+        joinNode->sources()[0]->outputType(),
+        joinNode->sources()[1]->outputType());
   }
 }
 
 BlockingReason NestedLoopJoinProbe::isBlocked(ContinueFuture* future) {
-  if (buildVectors_.has_value()) {
-    return BlockingReason::kNotBlocked;
-  }
+  switch (state_) {
+    case ProbeOperatorState::kRunning:
+      FOLLY_FALLTHROUGH;
+    case ProbeOperatorState::kFinish:
+      return BlockingReason::kNotBlocked;
+    case ProbeOperatorState::kWaitForBuild: {
+      VELOX_CHECK(!buildVectors_.has_value());
+      if (!getBuildData(future)) {
+        return BlockingReason::kWaitForJoinBuild;
+      }
+      VELOX_CHECK(buildVectors_.has_value());
 
-  auto buildVectors =
-      operatorCtx_->task()
-          ->getNestedLoopJoinBridge(
-              operatorCtx_->driverCtx()->splitGroupId, planNodeId())
-          ->dataOrFuture(future);
-  if (!buildVectors.has_value()) {
-    return BlockingReason::kWaitForJoinBuild;
-  }
-
-  buildVectors_ = std::move(buildVectors);
+      if (needsBuildMismatch(joinType_)) {
+        buildMatched_.resize(buildVectors_->size());
+        for (auto i = 0; i < buildVectors_->size(); ++i) {
+          buildMatched_[i].resizeFill(buildVectors_.value()[i]->size(), false);
+        }
+      }
 
-  if (buildVectors_->empty()) {
-    // Build side is empty. Return empty set of rows and  terminate the pipeline
-    // early.
-    buildSideEmpty_ = true;
+      setState(ProbeOperatorState::kRunning);
+      return BlockingReason::kNotBlocked;
+    }
+    case ProbeOperatorState::kWaitForPeers: {
+      if (future_.valid()) {
```

Comment:
We probably doesn't need future_ see comment below. thanks!

Reply:
Removed, thanks!